### ISSUE TITLE
NaNs for BF16 2D/3D convolution forward.

### ISSUE THREAD

**Issue by aosewski** [ROCmSoftwarePlatform/composable_kernel#135]
Issue in this PR #133

Steps to reproduce on ROCm 5.0 and 4.3.1

branch: [aosewski/conv_3d](https://github.com/ROCmSoftwarePlatform/composable_kernel/tree/aosewski/conv_3d)
The issue occures only for first instance of generic convolution algorithm. Like this one:
https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/d332ff7f3ab2e9e415ed516611c27cf6780cf7da/library/src/tensor_operation_instance/gpu/conv1d_fwd/device_conv1d_fwd_xdl_nwc_kxc_nwk_bf16_instance.cpp#L37
https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/d332ff7f3ab2e9e415ed516611c27cf6780cf7da/library/src/tensor_operation_instance/gpu/conv2d_fwd/device_conv2d_fwd_xdl_nhwc_kyxc_nhwk_bf16_instance.cpp#L38
https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/d332ff7f3ab2e9e415ed516611c27cf6780cf7da/library/src/tensor_operation_instance/gpu/conv3d_fwd/device_conv3d_fwd_xdl_ndhwc_kzyxc_ndhwk_bf16_instance.cpp#L37
So need to uncomment those lines.

The issue happens only when initializing input tensors with integer values. Thus one need to change current input/weights tensor initialization to integer values initialization. You can use
[init.diff.txt](https://github.com/ROCmSoftwarePlatform/composable_kernel/files/8284448/init.diff.txt) file.

```
cmake -DBUILD_DEV=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS="-DCK_AMD_GPU_GFX908 --amdgpu-target=gfx908 -O3" -DCMAKE_CXX_COMPILER=/opt/rocm/bin/hipcc -DCMAKE_PREFIX_PATH="/opt/rocm/;/opt/rocm/hip" ..
cmake --build . -j 36 --target test_convnd_fwd
./bin/test_conv1d_fwd
./bin/test_conv2d_fwd
./bin/test_conv3d_fwd
```

The error looks like following:
```
DeviceConv1DFwdXdl_Input_N_Hi_Wi_C_Weight_K_Y_X_C_Output_N_Ho_Wo_K<256, 256, 128, 4, Default>....                Supported!
launch_and_time_kernel: grid_dim {36, 1, 1}, block_dim {256, 1, 1}
Warm up
Start running 1 times...
        out[0] != ref[0]: -nan != 0
Error: incorrect results!
        out[1] != ref[1]: -nan != 0
Error: incorrect results!
        out[2] != ref[2]: -nan != 0
Error: incorrect results!
        out[3] != ref[3]: -nan != 0
Error: incorrect results!
   max err: 0

TestConv1DNWCBF16Instances ..... FAILURE
>>>>>>>> RUN test conv2d nhwc <<<<<<<<<<
DeviceConv2DFwdXdl_Input_N_Hi_Wi_C_Weight_K_Y_X_C_Output_N_Ho_Wo_K<256, 256, 128, 4, Default>....                Supported!
launch_and_time_kernel: grid_dim {1296, 1, 1}, block_dim {256, 1, 1}
Warm up
Start running 1 times...
        out[0] != ref[0]: -nan != -nan
Error: incorrect results!
        out[1] != ref[1]: -nan != -nan
Error: incorrect results!
        out[2] != ref[2]: -nan != -nan
Error: incorrect results!
        out[3] != ref[3]: -nan != -nan
Error: incorrect results!
   max err: 0

TestConv2DNHWCBF16Instances ..... FAILURE
>>>>>>>> RUN test conv3d ndhwc <<<<<<<<<<
DeviceConv3DFwdXdl_Input_N_Hi_Wi_C_Weight_K_Y_X_C_Output_N_Ho_Wo_K<256, 256, 128, 4, Default>....                Supported!
launch_and_time_kernel: grid_dim {256, 1, 1}, block_dim {256, 1, 1}
Warm up
Start running 1 times...
        out[0] != ref[0]: -nan != -nan
Error: incorrect results!
        out[1] != ref[1]: -nan != -nan
Error: incorrect results!
        out[2] != ref[2]: -nan != -nan
Error: incorrect results!
        out[3] != ref[3]: -nan != -nan
Error: incorrect results!
   max err: 0

TestConv3DNDHWCBF16Instances ..... FAILURE
```

**Comment by asroy** [comment]
@aosewski Does this issue still exsit?

**Comment by aosewski** [comment]
@asroy Yes

**Comment by aosewski** [comment]
|  rocm  |  commit#  |  branch  |  init  |  conv1D Bf16 |  conv2d Bf16  |  conv3D Bf16   |  conv config  |
| :---:  |  :---:    |  :---:   | :---:  |  :---:       |  :---:        |  :---:         |  :---:        |
|  4.3.1 |  4e6dfda  |  conv_3d |  dec   |-8.375 != -nan|               | max err: 0.25  | <256, 256, 128, 4, Default> |
|  4.3.1 |  2868f2e  |  conv_3d |  int   |  -nan != 0   |  -nan != -nan | -nan != -nan   | <256, 256, 128, 4, Default> |
|  4.3.1 |  9e33fe7  |  develop |  dec   |   NA         | max err: 0.25 |     NA         | all supported |
|  4.3.1 |  9e33fe7  |  develop |  int   |   NA         | OK            |     NA         | all supported |
|  5.0   |  9e33fe7  |  develop |  dec   |   NA         | max err: 0.25 |     NA         | all supported |
|  5.0   |  9e33fe7  |  develop |  int   |   NA         | OK            |     NA         | all supported |
|  5.0   |  4e6dfda  |  conv_3d |  int   | -nan != 0    | -nan != -nan  | -nan != -nan   | <256, 256, 128, 4, Default> |
|  5.0   |  4e6dfda  |  conv_3d |  dec   | OK           | OK            |  OK            | <256, 256, 128, 4, Default> |
|  5.0   |  2868f2e  |  conv_3d |  int   | -nan != 0    | -nan != -nan  | -nan != -nan   | <256, 256, 128, 4, Default> |

**Comment by aosewski** [comment]
@asroy Should be fixed with #259 

The failing instance is now tested in unit tests: 
https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/7a4bd7bc6bee74ea7500f24fd5ced314460798db/test/convnd_fwd/conv2d_fwd.cpp#L224-L227
https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/7a4bd7bc6bee74ea7500f24fd5ced314460798db/library/src/tensor_operation_instance/gpu/conv2d_fwd/device_convnd_2d_fwd_xdl_nhwc_kyxc_nhwk_bf16_instance.cpp#L37
![ck_private_ndbf16_default_nan](https://user-images.githubusercontent.com/19374865/171901218-ac3b4f28-6c86-4fd9-b924-4fbf1fa63cb8.png)

Tested on MI100 under: 
* compute-artifactory.amd.com:5000/rocm-plus-docker/framework/compute-rocm-dkms-no-npi-hipclang:9110_ubuntu18.04_py3.6_pytorch_rocm5.0_internal_testing_7ff5b54
* rocm/ck-private:9110_ubuntu18.04_py3.6_pytorch
* rocm/dev-ubuntu-20.04:5.1.1-complete

The issue is no longer observed.

**Comment by asroy** [comment]
fixed
